### PR TITLE
Fix retries and extend testing

### PR DESCRIPTION
**Pull request by chrimc62**
Retries was not properly reset on all memory modifications.  Added unit tests to verify the fixes which required adding MemoryAssertions/SetProperties to test scripts.  Also removed the existing property mocking since it was less useful than SetProperties.

### CODE REVIEW

**Review by coldplaying42 — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SetProperties.cs [hunk 1]
```diff
@@ -81,6 +81,9 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
                 dc.State.SetValue(propValue.Property.GetValue(dc.State), value);
             }
 
+            // Explicit state change resets retries
+            dc.State.RemoveValue(DialogPath.Retries);
```
Comment (mrivera-ms) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SetProperties.cs:85]: This reset in retries happens regardless of having a change in state. Should we add a flag inside the 'foreach' to indicate this change needs to be made after the loop?

Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SetProperties.cs:85]: Just calling this is enough to reset even if nothing changed about the state. Retries is used to identify when the same conceptual query is asked multiple times.  Any memory setting indicates a response was processed and we should reset it.

---
In reply to: [490538953](https://github.com/microsoft/botbuilder-dotnet/pull/4665#discussion_r490538953) [](ancestors = 490538953)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/DialogInspector.cs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            await dialogStateManager.LoadAllScopesAsync(cancellationToken).ConfigureAwait(false);
+            dc.Context.TurnState.Add(dialogStateManager);
+
+            // Find the DC for the active dialog
+            var activeDc = GetActiveDialogContext(dc);
+
+            // save all state scopes to their respective botState locations.
+            await dialogStateManager.SaveAllChangesAsync(cancellationToken).ConfigureAwait(false);
+
+            inspector(activeDc);
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/DialogInspector.cs:201]: >inspector(activeDc); [](start = 12, length = 20)

Why is this between these two save state calls? The call above is pointless as nothing would have changed as we didn't call any dialogs.

**Review by Stevenic — APPROVED** [review]
:shipit:

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/DialogInspector.cs (hunk 2 above)
Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/DialogInspector.cs]: Fixed per suggestion.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/AdaptiveTestingComponentRegistration.cs [hunk 3]
```diff
@@ -36,9 +39,10 @@ public virtual IEnumerable<DeclarativeType> GetDeclarativeTypes(ResourceExplorer
             yield return new DeclarativeType<AssertReply>(AssertReply.Kind);
             yield return new DeclarativeType<AssertReplyOneOf>(AssertReplyOneOf.Kind);
             yield return new DeclarativeType<AssertReplyActivity>(AssertReplyActivity.Kind);
+            yield return new DeclarativeType<MemoryAssertions>(MemoryAssertions.Kind);
             yield return new DeclarativeType<HttpRequestSequenceMock>(HttpRequestSequenceMock.Kind);
             yield return new DeclarativeType<UserTokenBasicMock>(UserTokenBasicMock.Kind);
-            yield return new DeclarativeType<PropertiesMock>(PropertiesMock.Kind);
+            yield return new DeclarativeType<SetProperties>(SetProperties.Kind);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/AdaptiveTestingComponentRegistration.cs:45]: >pe<SetProperties>(SetProperties.Kind); [](start = 42, length = 38)

Wha? ???

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/AssertReplyActivity.cs [hunk 4]
```diff
@@ -84,9 +85,9 @@ public virtual void ValidateReply(Activity activity)
         }
 
         /// <inheritdoc/>
-        public override async Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler callback)
+        public override async Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler callback, DialogInspector inspector)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/AssertReplyActivity.cs:88]: >k, DialogInspector inspector) [](start = 95, length = 29)

This is required and not used?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/CustomEvent.cs [hunk 5]
```diff
@@ -56,9 +53,9 @@ public CustomEvent([CallerFilePath] string path = "", [CallerLineNumber] int lin
         public object Value { get; set; }
 
         /// <inheritdoc/>
-        public async override Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler callback)
+        public async override Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler callback, DialogInspector inspector)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/CustomEvent.cs:56]: >, DialogInspector inspector [](start = 96, length = 27)

this is required and not used?

**Pull request merged**